File: apps/api/app/response_security.py
from __future__ import annotations

from app.runtime import AppSettings

BASELINE_RESPONSE_HEADERS = {
    "Referrer-Policy": "strict-origin-when-cross-origin",
    "X-Content-Type-Options": "nosniff",
    "X-Frame-Options": "DENY",
    "Permissions-Policy": "camera=(), geolocation=(), microphone=(), payment=()",
}
STRICT_TRANSPORT_SECURITY_VALUE = "max-age=63072000; includeSubDomains; preload"
EXPERIMENTAL_API_PREFIXES = (
    "/v1/auth",
    "/v1/profiles",
    "/v1/consents",
    "/v1/upload-sessions",
)
EXPERIMENTAL_API_PATHS = {
    "/v1/research-pulse/for-you",
}


def is_experimental_api_path(path: str) -> bool:
    normalized_path = path.rstrip("/") or "/"
    if normalized_path in EXPERIMENTAL_API_PATHS:
        return True
    return any(normalized_path.startswith(prefix) for prefix in EXPERIMENTAL_API_PREFIXES)


def build_response_security_headers(*, path: str, settings: AppSettings) -> dict[str, str]:
    headers = dict(BASELINE_RESPONSE_HEADERS)

    if settings.env_name in {"staging", "production"}:
        headers["Strict-Transport-Security"] = STRICT_TRANSPORT_SECURITY_VALUE

    if is_experimental_api_path(path):
        headers["Cache-Control"] = "private, no-store, max-age=0"
        headers["X-OneRhythm-Release-Stage"] = "experimental"

    return headers

File: apps/web/lib/response-security.ts
type ResponseHeader = {
  key: string;
  value: string;
};

type HeaderRule = {
  source: string;
  headers: ResponseHeader[];
};

const BASELINE_HEADERS: ResponseHeader[] = [
  { key: "Referrer-Policy", value: "strict-origin-when-cross-origin" },
  { key: "X-Content-Type-Options", value: "nosniff" },
  { key: "X-Frame-Options", value: "DENY" },
  { key: "Permissions-Policy", value: "camera=(), geolocation=(), microphone=(), payment=()" },
];

const HSTS_HEADER: ResponseHeader = {
  key: "Strict-Transport-Security",
  value: "max-age=63072000; includeSubDomains; preload",
};

const EXPERIMENTAL_HEADERS: ResponseHeader[] = [
  { key: "Cache-Control", value: "private, no-store, max-age=0" },
  { key: "X-OneRhythm-Release-Stage", value: "experimental" },
  { key: "X-Robots-Tag", value: "noindex, nofollow, noarchive" },
];

export const experimentalRouteSources = [
  "/sign-in",
  "/sign-up",
  "/onboarding/:path*",
  "/education/:path*",
  "/account/:path*",
] as const;

export function buildWebResponseHeaderRules(environment: string = process.env.NODE_ENV ?? "development"): HeaderRule[] {
  const normalizedEnvironment = environment.trim().toLowerCase();
  const baselineHeaders =
    normalizedEnvironment === "production" || normalizedEnvironment === "staging"
      ? [...BASELINE_HEADERS, HSTS_HEADER]
      : BASELINE_HEADERS;

  return [
    {
      source: "/:path*",
      headers: baselineHeaders,
    },
    ...experimentalRouteSources.map((source) => ({
      source,
      headers: EXPERIMENTAL_HEADERS,
    })),
  ];
}

